File: kelvinWu/reverseLinked/lc25/lc25.js
// Problem Statement #

// Given the head of a LinkedList and a number ‘k’, reverse every ‘k’ sized sub-list starting from the head.

// If, in the end, you are left with a sub-list with less than ‘k’ elements, reverse it too.

class Node {
  constructor(value, next = null) {
    this.value = value;
    this.next = next;
  }

  get_list() {
    let result = "";
    let temp = this;
    while (temp !== null) {
      result += temp.value + " ";
      temp = temp.next;
    }
    return result;
  }
}

// k = 3
// s     e
// 1 2 3 4 5 6
// 1 2 3
//               s     e
//        3 2 1  4 5 6 null

// 3 2 1 6 5 4

// 3 2 1 6 5 4

//loop through linkedlist
////in the loop, reverse linkedlist up to the node k ahead
////add reverse sublist to new linkedlist
//return new linkedlist

const reverse_every_k_elements = function (head, k) {
  // TODO: Write your code here
  let dummyNode = new Node(null);
  let newHead = dummyNode;
  let currentHead = head;

  while (currentHead !== null) {
    let prev = null;
    let current = currentHead;
    let subListSize = 0;
    while (current !== null && subListSize < k) {
      let temp = current.next;
      current.next = prev;
      prev = current;
      current = temp;
      subListSize++;
    }
    dummyNode.next = prev;
    dummyNode = currentHead;
    currentHead = current;
  }

  return newHead.next;
};
// k = 3
// 1 2 3 4 5 6
// currentHead = 1
// size = 0    prev = 1  current = 2
// size = 1   prev = 2 current = 3
// size = 2 prev = 3 current = 4
// dummyNode.next = 3
// currentHead = 4

// function reverse_every_k_elements(head, k) {
//   if (k <= 1 || head === null) {
//     return head;
//   }

//   let current = head,
//     previous = null;
//   while (true) {
//     const last_node_of_previous_part = previous;
//     // after reversing the LinkedList 'current' will become the last node of the sub-list
//     const last_node_of_sub_list = current;
//     let next = null; // will be used to temporarily store the next node
//     let i = 0;
//     while (current !== null && i < k) { // reverse 'k' nodes
//       next = current.next;
//       current.next = previous;
//       previous = current;
//       current = next;
//       i += 1;
//     }

//     // connect with the previous part
//     if (last_node_of_previous_part !== null) {
//       last_node_of_previous_part.next = previous;
//     } else {
//       head = previous;
//     }
//     // connect with the next part
//     last_node_of_sub_list.next = current;

//     if (current === null) {
//       break;
//     }
//     previous = last_node_of_sub_list;
//   }
//   return head;
// }

head = new Node(1);
head.next = new Node(2);
head.next.next = new Node(3);
head.next.next.next = new Node(4);
head.next.next.next.next = new Node(5);
head.next.next.next.next.next = new Node(6);
head.next.next.next.next.next.next = new Node(7);
head.next.next.next.next.next.next.next = new Node(8);

console.log(`Nodes of original LinkedList are: ${head.get_list()}`);
console.log(
  `Nodes of reversed LinkedList are: ${reverse_every_k_elements(
    head,
    3
  ).get_list()}`
);

// Solution
// -----
// function reverse_every_k_elements(head, k) {
//   if (k <= 1 || head === null) {
//     return head;
//   }

//   let current = head,
//     previous = null;
//   while (true) {
//     const last_node_of_previous_part = previous;
//     // after reversing the LinkedList 'current' will become the last node of the sub-list
//     const last_node_of_sub_list = current;
//     let next = null; // will be used to temporarily store the next node
//     let i = 0;
//     while (current !== null && i < k) { // reverse 'k' nodes
//       next = current.next;
//       current.next = previous;
//       previous = current;
//       current = next;
//       i += 1;
//     }

//     // connect with the previous part
//     if (last_node_of_previous_part !== null) {
//       last_node_of_previous_part.next = previous;
//     } else {
//       head = previous;
//     }
//     // connect with the next part
//     last_node_of_sub_list.next = current;

//     if (current === null) {
//       break;
//     }
//     previous = last_node_of_sub_list;
//   }
//   return head;
// }

// -----

// Time complexity #
// The time complexity of our algorithm will be O(N) where ‘N’ is the total number of nodes in the LinkedList.

// Space complexity #
// We only used constant space, therefore, the space complexity of our algorithm is O(1).

File: kelvinWu/reverseLinked/lc25/lc25.cpp
// Problem Statement #

// Given the head of a LinkedList and a number ‘k’, reverse every ‘k’ sized sub-list starting from the head.

// If, in the end, you are left with a sub-list with less than ‘k’ elements, reverse it too.

using namespace std;

#include <iostream>

class ListNode {
 public:
  int value = 0;
  ListNode *next;

  ListNode(int value) {
    this->value = value;
    next = nullptr;
  }
};

class ReverseEveryKElements {
 public:
  static ListNode *reverse(ListNode *head, int k) {
    // TODO: Write your code here
    return head;
  }
};

int main(int argc, char *argv[]) {
  ListNode *head = new ListNode(1);
  head->next = new ListNode(2);
  head->next->next = new ListNode(3);
  head->next->next->next = new ListNode(4);
  head->next->next->next->next = new ListNode(5);
  head->next->next->next->next->next = new ListNode(6);
  head->next->next->next->next->next->next = new ListNode(7);
  head->next->next->next->next->next->next->next = new ListNode(8);

  ListNode *result = ReverseEveryKElements::reverse(head, 3);
  cout << "Nodes of the reversed LinkedList are: ";
  while (result != nullptr) {
    cout << result->value << " ";
    result = result->next;
  }
}

// Solution
// -----
// class ReverseEveryKElements {
//  public:
//   static ListNode *reverse(ListNode *head, int k) {
//     if (k <= 1 || head == nullptr) {
//       return head;
//     }

//     ListNode *current = head, *previous = nullptr;
//     while (true) {
//       ListNode *lastNodeOfPreviousPart = previous;
//       // after reversing the LinkedList 'current' will become the last node of the sub-list
//       ListNode *lastNodeOfSubList = current;
//       ListNode *next = nullptr;  // will be used to temporarily store the next node
//       // reverse 'k' nodes
//       for (int i = 0; current != nullptr && i < k; i++) {
//         next = current->next;
//         current->next = previous;
//         previous = current;
//         current = next;
//       }

//       // connect with the previous part
//       if (lastNodeOfPreviousPart != nullptr) {
//         lastNodeOfPreviousPart->next =
//             previous;  // 'previous' is now the first node of the sub-list
//       } else {         // this means we are changing the first node (head) of the LinkedList
//         head = previous;
//       }

//       // connect with the next part
//       lastNodeOfSubList->next = current;

//       if (current == nullptr) {  // break, if we've reached the end of the LinkedList
//         break;
//       }
//       // prepare for the next sub-list
//       previous = lastNodeOfSubList;
//     }

//     return head;
//   }
// };

// -----

// Time complexity #
// The time complexity of our algorithm will be O(N) where ‘N’ is the total number of nodes in the LinkedList.

// Space complexity #
// We only used constant space, therefore, the space complexity of our algorithm is O(1).
